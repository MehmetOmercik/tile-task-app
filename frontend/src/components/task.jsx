// import React from "react";
import PropTypes from "prop-types";
import { capitalise } from "../utils/helpers";

export const Task = ({
  id = "null",
  title = "null",
  order = "null",
  description = "null",
  type = "null",
}) => {
  return (
    <section className="border border-gray-600 rounded-md w-[200px] p-10 text-center text-lg bg-white">
      <p>Task {id}</p>
      <p>{title}</p>
      <p>{order}</p>
      <p>{description}</p>
      <p>{capitalise(type)}</p>
    </section>
  );
};

Task.propTypes = {
  id: PropTypes.number,
  title: PropTypes.string,
  order: PropTypes.number,
  description: PropTypes.string,
  type: PropTypes.string,
};
