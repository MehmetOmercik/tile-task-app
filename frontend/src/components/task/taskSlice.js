import { createSlice } from "@reduxjs/toolkit";

const taskInitialState = {
  editTask: false,
  editTaskObject: {
    title: "",
    description: "",
    order: 0,
    type: "survey",
    tile: 0,
  },
};

export const taskSlice = createSlice({
  name: "tile",
  initialState: taskInitialState,
  reducers: {
    updateEditTask: (state, action) => {
      state.editTask = action.payload;
    },
    updateEditTaskObject: (state, action) => {
      switch (action.payload.type) {
        case "UPDATE_PROPERTY":
          return {
            ...state,
            editTaskObject: {
              ...state.editTaskObject,
              [action.payload.property]: action.payload.value,
            },
          };
        case "UPDATE_OBJECT":
          return {
            ...state,
            editTaskObject: action.payload.payload,
          };
      }
    },
  },
});

export const { updateEditTask, updateEditTaskObject } = taskSlice.actions;

export default taskSlice.reducer;
