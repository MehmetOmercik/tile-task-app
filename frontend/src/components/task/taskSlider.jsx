import { Task } from "./task";
import { useState } from "react";
export const TaskSlider = ({ tasks }) => {
  const [currentIndex, setCurrentIndex] = useState(0);

  const goToPrevious = () => {
    const isFirstSlide = currentIndex === 0;
    const newIndex = isFirstSlide ? tasks.length - 1 : currentIndex - 1;
    setCurrentIndex(newIndex);
  };
  const goToNext = () => {
    const isLastSlide = currentIndex === tasks.length - 1;
    const newIndex = isLastSlide ? 0 : currentIndex + 1;
    setCurrentIndex(newIndex);
  };
  const goToSlide = (slideIndex) => {
    setCurrentIndex(slideIndex);
  };

  const taskIndex = tasks[currentIndex];

  return (
    <section>
      {tasks.length > 0 ? (
        <div className="flex items-center gap-x-4">
          {tasks.length > 1 && (
            <div
              onClick={goToPrevious}
              className="text-4xl cursor-pointer hover:text-blue-500"
            >
              ❰
            </div>
          )}

          <Task
            key={`Task ${taskIndex?.id}`}
            id={taskIndex?.id}
            title={taskIndex?.title}
            order={taskIndex?.order}
            description={taskIndex?.description}
            type={taskIndex?.type}
          />

          {tasks.length > 1 && (
            <div
              onClick={goToNext}
              className="text-4xl cursor-pointer hover:text-blue-500"
            >
              ❱
            </div>
          )}
        </div>
      ) : (
        <p className="h-[200px] w-[200px] text-center text-6xl">
          No tasks found
        </p>
      )}
      <div className="flex justify-center">
        {tasks.map((_, slideIndex) => (
          <div
            className={`mx-3 mt-3 cursor-pointer text-2xl hover:scale-125 hover:text-blue-500
                ${currentIndex === slideIndex ? " text-green-500" : ""}`}
            key={slideIndex}
            onClick={() => goToSlide(slideIndex)}
          >
            ●
          </div>
        ))}
      </div>
    </section>
  );
};
