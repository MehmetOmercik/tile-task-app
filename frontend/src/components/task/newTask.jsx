import { useDispatch, useSelector } from "react-redux";
import { updateEditTaskObject } from "./taskSlice";
import { DropdownComponent } from "../dropdownComponent";
import { putTask, postTask } from "../../utils/http";

export const NewTask = ({ setOverlaySection }) => {
  const dispatch = useDispatch();
  const { editTask: editTaskBoolean, editTaskObject } = useSelector(
    (state) => state.task
  );
  const { currentTileID } = useSelector((state) => state.tile);

  const handleProperty = (e, property) => {
    const value = e.target.value;

    dispatch(
      updateEditTaskObject({
        type: "UPDATE_PROPERTY",
        property,
        value,
      })
    );
  };

  const handleTypeOptionSelect = (selectedOption) => {
    console.log(selectedOption);
    dispatch(
      updateEditTaskObject({
        type: "UPDATE_PROPERTY",
        property: "type",
        value: selectedOption,
      })
    );
  };

  // const handleTileOptionSelect = (selectedOption) => {
  //   console.log(selectedOption);
  //   dispatch(
  //     updateEditTaskObject({
  //       type: "UPDATE_PROPERTY",
  //       property: "tile",
  //       value: selectedOption,
  //     })
  //   );
  // };

  const handleTaskSubmission = async (e) => {
    const taskPayload = {
      // id: editTaskObject.id
      title: editTaskObject.title,
      description: editTaskObject.description,
      order: +editTaskObject.order,
      type: editTaskObject.type,
      tile: currentTileID,
    };

    // e.preventDefault();
    editTaskBoolean
      ? putTask(editTaskObject.id, taskPayload)
      : postTask(taskPayload);

    // ! Doesnt update properly
    // window.location.reload();
    // await setOverlaySection("tasks");
  };

  return (
    <section className="bg-white  w-[300px] rounded-xl p-4">
      <h1 className="text-2xl mb-5">
        {editTaskBoolean
          ? `Update Task ${editTaskObject.id} for Tile ${currentTileID}`
          : `Create Task for Tile ${currentTileID}`}
      </h1>
      <form onSubmit={(e) => handleTaskSubmission(e)}>
        <label>Title</label>
        <input
          maxLength="30"
          value={editTaskObject.title}
          onChange={(e) => handleProperty(e, "title")}
          className="bg-gray-200 rounded-lg p-1 w-[97%]"
        ></input>

        <label htmlFor="">Description</label>
        <textarea
          value={editTaskObject.description}
          onChange={(e) => handleProperty(e, "description")}
          className="resize-none bg-gray-200 rounded-lg h-20 w-[97%] p-1"
          maxLength="256"
        ></textarea>

        <label htmlFor="">Order</label>
        <input
          value={editTaskObject.order}
          onChange={(e) => handleProperty(e, "order")}
          className="bg-gray-200 rounded-lg p-1 w-[97%] "
        ></input>

        <label>Type</label>
        <DropdownComponent
          options={["survey", "discussion", "diary"]}
          onSelect={handleTypeOptionSelect}
          defaultValue={editTaskObject.type}
        />
        {/* <label>Tile</label>
        <DropdownComponent
          options={[1, 2, 3]}
          onSelect={handleTileOptionSelect}
          defaultValue={editTaskObject.tile}
        /> */}
        <button
          type="submit"
          disabled={!editTaskObject.type}
          className="bg-slate-400 rounded-lg p-2 text-lg hover:bg-slate-300 disabled:bg-red-600"
        >
          {editTaskBoolean ? "Update" : "Create"}
        </button>
      </form>
    </section>
  );
};
