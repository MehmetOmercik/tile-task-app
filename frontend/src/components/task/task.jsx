import trashCan from "../../assets/trashCan.svg";
import editIcon from "../../assets/editIcon.svg";
import PropTypes from "prop-types";
import { capitalise } from "../../utils/helpers";
import { deleteTask } from "../../utils/http";
import { useDispatch, useSelector } from "react-redux";
import { updateEditTask, updateEditTaskObject } from "./taskSlice";

export const Task = ({
  taskID = "null",
  title = "null",
  order = "null",
  description = "null",
  type = "null",
  setOverlaySection,
}) => {
  const dispatch = useDispatch();
  const { currentTileID } = useSelector((state) => state.tile);

  const handleEditTask = () => {
    const taskPayload = {
      id: taskID,
      title,
      order,
      description,
      type,
      tile: currentTileID,
    };
    setOverlaySection("edit_task");
    dispatch(updateEditTask(true));
    dispatch(
      updateEditTaskObject({ type: "UPDATE_OBJECT", payload: taskPayload })
    );
  };

  const handleDeleteTask = async () => {
    await deleteTask(taskID);
    window.location.reload();
  };

  return (
    <section
      id="task-container"
      className="border border-gray-600 rounded-md w-[320px] text-center text-lg bg-white"
    >
      <div className="p-4">
        <h1 className="text-2xl font-bold break-words">{title}</h1>
      </div>
      <div className="border"></div>
      <h2 className="text-left m-2 font-semibold">Description</h2>
      <p className="w-full resize-none break-words p-2">{description}</p>
      <div className="flex justify-between p-2 font-semibold">
        <p>{capitalise(type)}</p>
        <p>Task ID: {taskID}</p>
        <p>Order NO: {order}</p>
      </div>
      <div className="flex justify-between px-2 mb-6">
        <img
          onClick={handleEditTask}
          className="rounded-lg hover:cursor-pointer hover:bg-gray-200 p-2"
          src={editIcon}
          width="60px"
        />
        <img
          onClick={handleDeleteTask}
          className="rounded-lg hover:cursor-pointer hover:bg-gray-200 p-2"
          src={trashCan}
          width="60px"
        />
      </div>
    </section>
  );
};

Task.propTypes = {
  taskID: PropTypes.number,
  title: PropTypes.string,
  order: PropTypes.number,
  description: PropTypes.string,
  type: PropTypes.string,
};
