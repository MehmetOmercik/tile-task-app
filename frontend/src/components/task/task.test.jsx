import { Task } from "./task";
import { renderWithProviders } from "../../test/test-utils";

describe("Task tests", () => {
  test("Snapshot test", () => {
    const task = renderWithProviders(<Task />);
    expect(task).toMatchSnapshot();
  });

  test("Properties test", () => {
    const task = renderWithProviders(
      <Task
        taskID={1}
        title="test title"
        order={1}
        description="test description"
        type="survey"
      />
    );
    const titleElement = task.getByTestId("task-title-h1");
    const descriptionElement = task.getByTestId("task-description-p");
    const orderElement = task.getByTestId("task-order-p");
    const idElement = task.getByTestId("task-taskID-p");
    const typeElement = task.getByTestId("task-type-p");

    expect(titleElement).toHaveTextContent("test title");
    expect(descriptionElement).toHaveTextContent("test description");
    expect(orderElement).toHaveTextContent(1);
    expect(idElement).toHaveTextContent(1);
    expect(typeElement).toHaveTextContent("Survey");
  });
});
