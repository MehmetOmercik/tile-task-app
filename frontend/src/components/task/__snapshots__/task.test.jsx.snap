// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task tests > Snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="border border-gray-600 rounded-md w-[320px] text-center text-lg bg-white"
        id="task-container"
      >
        <div
          class="p-4"
        >
          <h1
            class="text-2xl font-bold break-words"
            data-testid="task-title-h1"
          >
            null
          </h1>
        </div>
        <div
          class="border"
        />
        <h2
          class="text-left m-2 font-semibold"
        >
          Description
        </h2>
        <p
          class="w-full resize-none break-words p-2"
          data-testid="task-description-p"
        >
          null
        </p>
        <div
          class="flex justify-between p-2 font-semibold"
        >
          <p
            data-testid="task-type-p"
          >
            Null
          </p>
          <p
            data-testid="task-taskID-p"
          >
            Task ID: 
            null
          </p>
          <p
            data-testid="task-order-p"
          >
            Order NO: 
            null
          </p>
        </div>
        <div
          class="flex justify-between px-2 mb-6"
        >
          <img
            class="rounded-lg hover:cursor-pointer hover:bg-gray-200 p-2"
            src="/src/assets/editIcon.svg"
            width="60px"
          />
          <img
            class="rounded-lg hover:cursor-pointer hover:bg-gray-200 p-2"
            src="/src/assets/trashCan.svg"
            width="60px"
          />
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="border border-gray-600 rounded-md w-[320px] text-center text-lg bg-white"
      id="task-container"
    >
      <div
        class="p-4"
      >
        <h1
          class="text-2xl font-bold break-words"
          data-testid="task-title-h1"
        >
          null
        </h1>
      </div>
      <div
        class="border"
      />
      <h2
        class="text-left m-2 font-semibold"
      >
        Description
      </h2>
      <p
        class="w-full resize-none break-words p-2"
        data-testid="task-description-p"
      >
        null
      </p>
      <div
        class="flex justify-between p-2 font-semibold"
      >
        <p
          data-testid="task-type-p"
        >
          Null
        </p>
        <p
          data-testid="task-taskID-p"
        >
          Task ID: 
          null
        </p>
        <p
          data-testid="task-order-p"
        >
          Order NO: 
          null
        </p>
      </div>
      <div
        class="flex justify-between px-2 mb-6"
      >
        <img
          class="rounded-lg hover:cursor-pointer hover:bg-gray-200 p-2"
          src="/src/assets/editIcon.svg"
          width="60px"
        />
        <img
          class="rounded-lg hover:cursor-pointer hover:bg-gray-200 p-2"
          src="/src/assets/trashCan.svg"
          width="60px"
        />
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
